honeybadgermpc/mpc.py

```diff
@@ -92,47 +108,65 @@ def call_mixin(self, name, *args, **kwargs):
 
         point = EvalPoint(self.field, self.N, use_fft=False)
 
         # Create polynomial that reconstructs the shared value by evaluating at 0
-        opening = robust_reconstruct(
+        p, _ = await robust_reconstruct(
             share_buffer, self.field, self.N, t, point)
-        self._openings[shareid] = opening
-
-        p, _ = await opening
 
         # Return reconstructed point
         return p(self.field(0))
 
-    def open_share_array(self, sharearray):
+    async def open_share_array(self, sharearray):
         # Choose the shareid based on the order this is called
-        shareid = len(self._openings)
+        shareid = self._get_share_id()
 
         # Creates unique send function based on the share to open
-        def _send(j, o):
+        def _send(dest, o):
             (tag, share) = o
-            self.send(j, (tag, shareid, share))
+            self.send(dest, (tag, shareid, share))
 
         # Receive function from the respective queue for this node
         _recv = self._sharearray_buffers[shareid].get
 
         # Generate reconstructed array of shares
-        opening = batch_reconstruct([s.v for s in sharearray._shares],
-                                    self.field.modulus,
-                                    sharearray.t,
-                                    self.N,
-                                    self.myid,
-                                    _send,
-                                    _recv,
-                                    config=self.config.get(ConfigVars.Reconstruction),
-                                    debug=True)
-        self._openings[shareid] = opening
-        return opening
+        return await batch_reconstruct([s.v for s in sharearray._shares],
+                                       self.field.modulus,
+                                       sharearray.t,
+                                       self.N,
+                                       self.myid,
+                                       _send,
+                                       _recv,
+                                       config=self.config.get(ConfigVars.Reconstruction),
+                                       debug=True)
+
+    async def _run(self):
+        # Run receive loop as background task, until self.prog finishes
+        # Cancel the background task, even if there's an exception
+        bgtask = asyncio.create_task(self._recvloop())
+        result = asyncio.create_task(self.prog(self, **self.prog_args))
+        await asyncio.wait((bgtask, result), return_when=asyncio.FIRST_COMPLETED)
+
+        # bgtask should not exit early-- this should correspond to an error
+        if bgtask.done():
+            logging.error('Background task finished before prog')
+
+            bg_exception = bgtask.exception()
+            if not result.done():
+                result.cancel()
+
+            if bg_exception is None:
+                raise Exception('background task finished before prog!')
```

Comment:
`raise HoneyBadgerMpcException`, avoid raising generic exceptions!

Can you think of a case when this can happen?
As a rule, always log exceptions, we have seen things hang when there are exceptions but it is easier to spot a log.

Reply:
With the current `_recvloop`, no, but I think it's best to have this in here for completeness.